src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/ReferenceConfidenceModel.java

```diff
@@ -0,0 +1,547 @@
+package org.broadinstitute.hellbender.tools.walkers.haplotypecaller;
+
+import com.google.common.annotations.VisibleForTesting;
+import htsjdk.samtools.*;
+import htsjdk.samtools.util.Locatable;
+import htsjdk.variant.variantcontext.*;
+import htsjdk.variant.vcf.VCFHeaderLine;
+import htsjdk.variant.vcf.VCFSimpleHeaderLine;
+import org.apache.commons.math3.stat.descriptive.moment.Mean;
+import org.broadinstitute.hellbender.engine.AlignmentContext;
+import org.broadinstitute.hellbender.tools.walkers.genotyper.InfiniteRandomMatingPopulationModel;
+import org.broadinstitute.hellbender.tools.walkers.genotyper.PloidyModel;
+import org.broadinstitute.hellbender.utils.*;
+import org.broadinstitute.hellbender.utils.genotyper.ReadLikelihoods;
+import org.broadinstitute.hellbender.utils.genotyper.SampleList;
+import org.broadinstitute.hellbender.utils.haplotype.Haplotype;
+import org.broadinstitute.hellbender.utils.locusiterator.LocusIteratorByState;
+import org.broadinstitute.hellbender.utils.pileup.PileupElement;
+import org.broadinstitute.hellbender.utils.pileup.ReadPileup;
+import org.broadinstitute.hellbender.utils.read.AlignmentUtils;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.variant.GATKVCFConstants;
+import org.broadinstitute.hellbender.utils.variant.GATKVariantContextUtils;
+
+import java.io.File;
+import java.util.*;
+
+/**
+ * Code for estimating the reference confidence
+ *
+ * This code can estimate the probability that the data for a single sample is consistent with a
+ * well-determined REF/REF diploid genotype.
+ *
+ */
+public final class ReferenceConfidenceModel {
+
+    private final GenomeLocParser genomeLocParser;
+
+    private final SampleList samples;
+    private final int indelInformativeDepthIndelSize;
+
+    private static final boolean WRITE_DEBUGGING_BAM = false;
+    private final SAMFileWriter debuggingWriter;
+
+    /**
+     * Surrogate quality score for no base calls.
+     * <p>
+     * This is the quality assigned to deletion (so without its own base-call quality) pile-up elements,
+     * when assessing the confidence on the hom-ref call at that site.
+     * </p>
+     */
+    private static final byte REF_MODEL_DELETION_QUAL = 30;
+
+    /**
+     * Base calls with quality threshold lower than this number won't be considered when assessing the
+     * confidence on the hom-ref call.
+     */
+    private static final byte BASE_QUAL_THRESHOLD = 6;
+
+    /**
+     * Only base calls with quality strictly greater than this constant,
+     * will be considered high quality if they are part of a soft-clip.
+     */
+    private static final byte HQ_BASE_QUALITY_SOFTCLIP_THRESHOLD = 28;
+
+    protected static final int MAX_N_INDEL_INFORMATIVE_READS = 40; // more than this is overkill because GQs are capped at 99 anyway
```

Comment:
We have learn that 40 is not overkill and in fact causes some problems. Please add a TODO pointing to the corresponding issue in GATK3 (perhaps you want to create a GATK4 issue for this). 
https://github.com/broadinstitute/gsa-unstable/issues/1108